chain/chain/src/resharding.rs

```diff
@@ -318,129 +321,183 @@ impl Chain {
             };
 
             // Commit the store update.
-            {
+            let commit_time = {
                 let histogram = RESHARDING_BATCH_COMMIT_TIME.with_label_values(&metrics_labels);
-                let _timer = histogram.start_timer();
+                let timer = histogram.start_timer();
                 store_update.commit()?;
-            }
+                timer.stop_and_record()
+            };
 
-            RESHARDING_BATCH_COUNT.with_label_values(&metrics_labels).inc();
+            let batch_count = {
+                let metric = RESHARDING_BATCH_COUNT.with_label_values(&metrics_labels);
+                metric.inc();
+                metric.get()
+            };
             RESHARDING_BATCH_SIZE.with_label_values(&metrics_labels).add(size as i64);
 
+            debug!(target: "resharding", ?shard_uid, "batch {batch_count} prepared in {prepare_time}s, applied in {apply_time}s and committed in {commit_time}s");
+
             // sleep between batches in order to throttle resharding and leave
             // some resource for the regular node operation
             std::thread::sleep(config.get().batch_delay.unsigned_abs());
         }
+        Ok(())
+    }
+
+    fn build_state_for_split_shards_impl(
+        resharding_request: ReshardingRequest,
+    ) -> Result<HashMap<ShardUId, StateRoot>, Error> {
+        let ReshardingRequest {
+            tries,
+            prev_hash,
+            prev_prev_hash,
+            shard_uid,
+            state_root,
+            next_epoch_shard_layout,
+            config,
+            handle,
+            on_demand,
+            ..
+        } = resharding_request;
+        tracing::debug!(target: "resharding", config=?config.get(), ?shard_uid, "build_state_for_split_shards_impl starting");
+
+        let shard_id = shard_uid.shard_id();
+        let new_shards = next_epoch_shard_layout
+            .get_children_shards_uids(shard_id)
+            .ok_or(Error::InvalidShardId(shard_id))?;
+        let mut state_roots: HashMap<_, _> =
+            new_shards.iter().map(|shard_uid| (*shard_uid, Trie::EMPTY_ROOT)).collect();
+
+        RESHARDING_STATUS
+            .with_label_values(&[&shard_uid.to_string()])
+            .set(ReshardingStatus::BuildingState.into());
+
+        // function to map account id to shard uid in range of child shards
+        let checked_account_id_to_shard_uid =
+            get_checked_account_id_to_shard_uid_fn(shard_uid, new_shards, next_epoch_shard_layout);
+
+        if on_demand {
+            // When resharding is triggered on demand no special iterator chaining is required
+            // because we fallback to use the tries stored on disk.
+            let trie: Trie = tries.get_trie_for_shard(shard_uid, state_root);
+
+            // However some gymnastics are required to make the iterator signature match the one required by
+            // `apply_resharding_batches`.
+            let mut iter = trie.disk_iter()?.map(|item| {
```

Comment:
This iterator is quite slow. 1-2 days on SSD archival. Building the batch is the issue:

```
batch 810 prepared in 9.300821688s, applied in 0.047881789s and committed in 0.241051493s shard_uid=s1.v1
```